python/pyiceberg/catalog/jdbc.py

```diff
@@ -0,0 +1,475 @@
+
+
+GET_NAMESPACE_PROPERTIES_SQL = f"SELECT {NAMESPACE_NAME} FROM {NAMESPACE_PROPERTIES_TABLE_NAME} WHERE {CATALOG_NAME} = %s AND {NAMESPACE_NAME} LIKE %s LIMIT 1"
+GET_ALL_NAMESPACE_PROPERTIES_SQL = f"SELECT * FROM {NAMESPACE_PROPERTIES_TABLE_NAME} WHERE {CATALOG_NAME} = %s AND {NAMESPACE_NAME} = %s"
+
+# Custom SQL not from JDBCCatalog.java
+LIST_ALL_NAMESPACES_SQL = f"""
+SELECT DISTINCT ns FROM
+(
+    SELECT {TABLE_NAMESPACE} AS ns FROM {CATALOG_TABLE_NAME} 
+    WHERE {CATALOG_NAME} = %s 
+    UNION
+    SELECT {NAMESPACE_NAME} AS ns FROM {NAMESPACE_PROPERTIES_TABLE_NAME} 
+    WHERE {CATALOG_NAME} = %s
+) AS all_catalog_namespaces
+"""
+
+def _sqlite(**properties: str) -> Any:
+    parsed_uri = urlparse(properties.get("uri"))
+    return sqlite3.connect(database=parsed_uri.path, uri=False)
+
+def _postgresql(**properties: str) -> Any:
+    parsed_uri = urlparse(properties.get("uri"))
+    postgresql_props = {
+        "user": parsed_uri.username,
+        "password": parsed_uri.password,
+        "dbname": parsed_uri.path[1:],
+        "host": parsed_uri.hostname,
+        "port": parsed_uri.port,
+    }
+
+    return db.connect(**postgresql_props)
+
+SCHEME_TO_DB = {
+    "file": _sqlite,
+    "postgresql": _postgresql,
+}
+
+class JDBCCatalog(Catalog):
+    def __init__(self, name: str, **properties: str):
+        super().__init__(name, **properties)
+
+        # Get a database connection for a specific scheme.
+        uri = urlparse(self.properties.get("uri"))
+        uri_scheme = str(uri.scheme)
+        if uri_scheme not in SCHEME_TO_DB:
+            raise ValueError(f"No registered database for scheme: {uri_scheme}")
+        self._get_db_connection = SCHEME_TO_DB[uri_scheme]
+
+    def initialize_catalog_tables(self) -> None:
+        with self._get_db_connection(**self.properties) as conn:
+            try:
+                curs = conn.cursor()
+                curs.execute(CREATE_CATALOG_TABLE)
+                curs.execute(CREATE_NAMESPACE_PROPERTIES_TABLE)
+                conn.commit()
+            finally:
+                curs.close()
+
+
+    def _convert_jdbc_to_iceberg(self, jdbc_row: Dict[str, str]) -> Table:
+        # properties: Dict[str, str] = table.parameters
+        # if TABLE_TYPE not in properties:
+        #     raise NoSuchTableError(f"Property table_type missing, could not determine type: {table.dbName}.{table.tableName}")
+
+        # table_type = properties[TABLE_TYPE]
+        # if table_type.lower() != ICEBERG:
+        #     raise NoSuchIcebergTableError(
+        #         f"Property table_type is {table_type}, expected {ICEBERG}: {table.dbName}.{table.tableName}"
+        #     )
+
+        # if prop_metadata_location := properties.get(METADATA_LOCATION):
+        #     metadata_location = prop_metadata_location
+        # else:
+        #     raise NoSuchTableError(f"Table property {METADATA_LOCATION} is missing")
+        metadata_location = jdbc_row[METADATA_LOCATION]
```

Comment:
Probably we want to have a better error message here as commented out on the line above